### PR TITLE
Webservices as a standalone application Take 2

### PR DESCRIPTION
**Pull request by wilsonge**
This was a continuation of the work done in #13

This improves the abstraction between the soap and HAL classes by making a common `Joomla\Webservices\Webservices` class and attempts to try and break the integrations with Joomla out into the `Joomla\Webservices\Integrations` namespace so we can make future compatibility with Joomla 4 "simpler"

Large Item Todos:
- [ ] Refactor away from Legacy MVC to some sort of new MVC
- [ ] Decide if we want a JTable replacement or whether we want to drop everything into models (me and @mbabker have discussed this a few times but nothing really ever came out of it)
- [x] Implement a session for enqueing messages into the application. Joomla's session v1 is pretty much unusable. v2 is got finished by @mbabker today: https://github.com/joomla-framework/session/pull/12 but isn't merged yet so I didn't bother
- [ ] Routing (because I couldn't work out at a glance how the current routing system really worked)
- [ ] Work out how the component is going to interact with this app (99% of the CMS classes are not going to be injectable into this thing)
- [ ] Work out how the app interacts with the frontend and backend of Joomla (things like `JApplicationCms::isAdmin` or the `JPATH_SITE`/`JPATH_ADMIN` constants
- [ ] JLayouts don't exist on the framework level so decide if we want to port them into our app or use something like Twig, Plates, Blade or Moustache through the framework Renderer (currently I have ported JLayouts over)
- [ ] Deal with replacing JDocument (cause it's never gonna exist in the framework cause it's a CMS weak link as it is) - as a temporary measure I've ported a simplified version of it into the app structure
- [ ] Structure of the app - currently I only have media + index.php in the root. We might want to rework this in the future as a lot of cheap hosts don't allow this (which is why Joomla is installed by default totally into the root)

Specific Todos:
- [ ] JUri::base() has a dependency on the Joomla config object. When this is now a service provider orientated object suddenly things become really hard really fast because you've got to pass either the configuration or Container in and things don't work. Michael in his Joomla Next! Repo ended up adding in a setBase and setRoot methods (https://github.com/mbabker/joomla-next/blob/master/libraries/core/Uri/Uri.php#L123) - but it doesn't seem ideal either. For now I've removed that section and just used the autodetection but it needs to be fixed before we put things live.
- [ ] Do we want an interface for changing the application config (currently things that were in the params of the system plugin

### CODE REVIEW

File: src/Renderer/Base.php [hunk 1]
```diff
@@ -0,0 +1,628 @@
...
+use Joomla\Webservices\Application;
+
+use Joomla\DI\Container;
+use Joomla\DI\ContainerAwareTrait;
+use Joomla\DI\ContainerAwareInterface;
+
+/**
+ * Document class, provides an easy interface to parse and display a document
+ *
+ * @since  __DELPOY_VERSION__
```
Comment (chrisdavenport) [inline src/Renderer/Base.php:21]: Typo: should be **DEPLOY_VERSION**

File: src/Renderer/Hal.php [hunk 2]
```diff
@@ -17,55 +23,57 @@
...
-	 * @var    string
-	 * @since  1.2
-	 */
-	protected $name = 'joomla';
-
-	/**
 	 * Render all hrefs as absolute, relative is default
+	 *
+	 * @var    boolean
+	 * @since  __DELPOY_VERSION__
```
Comment (chrisdavenport) [inline src/Renderer/Hal.php:32]: Typo: should be **DEPLOY_VERSION**

File: src/Renderer/Hal.php [hunk 3]
```diff
@@ -17,55 +23,57 @@
...
 		// Set default mime type.
-		$this->_mime = 'application/hal+' . $this->documentFormat;
+		$this->setMimeEncoding('application/hal+' . $this->documentFormat, false);
 
 		// Set document type.
-		$this->_type = 'hal+' . $this->documentFormat;
+		$this->setType('hal+' . $this->documentFormat);
 
 		// Set absolute/relative hrefs.
 		$this->absoluteHrefs = isset($options['absoluteHrefs']) ? $options['absoluteHrefs'] : true;
```
Comment (chrisdavenport) [inline src/Renderer/Hal.php]: Since the default is supposed to be relative, this should probably default to false.

File: src/Renderer/Hal.php [hunk 4]
```diff
@@ -222,13 +220,13 @@ protected function relToAbs($hal, $absoluteHrefs)
...
+		$uri = Uri::getInstance($href);
+
 		if ($absoluteHrefs && substr($href, 0, 1) == '/')
 		{
-			$href = rtrim(JUri::base(), '/') . $href;
+			$href = rtrim($uri->base(), '/') . $href;
 		}
 
-		$uri = JUri::getInstance($href);
-
```
Comment (chrisdavenport) [inline src/Renderer/Hal.php:231]: Although $href now contains the correct absolute URL at this point, it isn't used to update $uri and so this method fails to convert relative URLs to absolute URLs.

Maybe line 227 should be $uri = JUri::getInstance(rtrim($uri->base(), '/') . $href);

File: src/Renderer/Base.php (hunk 1 above)
Comment (wilsonge) [inline src/Renderer/Base.php:21]: It's not a typo - it's just the standard we happen to use in the framework e.g. https://github.com/joomla-framework/application/commit/185b68fe55537281385b637a90e11856ea01c8da

Comment (chrisdavenport) [inline src/Renderer/Base.php:21]: DEPLOY_VERSION I understand, but what does DELPOY_VERSION mean?

File: src/Renderer/Hal.php (hunk 3 above)
Comment (wilsonge) [inline src/Renderer/Hal.php]: I'd suggest you merge that into the current version and then I can merge staging into this

File: src/Renderer/Base.php (hunk 1 above)
Comment (wilsonge) [inline src/Renderer/Base.php:21]: Ahh :p yah :p let me go fix that :D

File: src/Renderer/Hal.php (hunk 2 above)
Comment (wilsonge) [inline src/Renderer/Hal.php:32]: Fixed

File: src/Renderer/Base.php (hunk 1 above)
Comment (wilsonge) [inline src/Renderer/Base.php:21]: Fixed

File: src/Renderer/Hal.php (hunk 3 above)
Comment (wilsonge) [inline src/Renderer/Hal.php]: Rebased this onto the develop branch - so this should be fixed now

File: src/Renderer/Hal.php (hunk 4 above)
Comment (wilsonge) [inline src/Renderer/Hal.php:231]: Good spot! Fixed!

**Comment by chrisdavenport** [comment]
I think this is ready to merge, but before I do it, can you restore the plugin code?  For the moment we still need to keep the plugin working otherwise the admin component doesn't work.  We can remove it again at a later stage.

Thanks @wilsonge.

**Comment by javigomez** [comment]
What will be the workflow now?

I mean, for the tests before I did:
- download Joomla
- Install Joomla https://github.com/joomla-projects/webservices/pull/17/files#diff-eccf834bcae4ac520d748d03c927336dR14
- install the webservices extension: https://github.com/joomla-projects/webservices/pull/17/files#diff-0c314fdf3b778869f6b2fc1e5417f90cR17 that includes the plugin
- Enable the plugin https://github.com/joomla-projects/webservices/pull/17/files#diff-210fd930d36f66f71e6e2f4fe78f5f8fR17
- Enable webservices in the plugin https://github.com/joomla-projects/webservices/pull/17/files#diff-210fd930d36f66f71e6e2f4fe78f5f8fR21
- Enable joomla basic authentication
- Install contact webservices
- Execute API tests

Now that is an app what would be the workflow for the tests?

**Comment by fastslack** [comment]
This is an initial aproach to new MVC administrator component for webservices. Any comments very welcome. I think i ll need some help integrating the layouts to new MVC.

https://github.com/fastslack/webservices/tree/newmvc/component/admin

**Comment by wilsonge** [comment]
It looks like a decent start - I'm assuming a lot of the display controller comes from Patchtester. One of your biggest issues right now is that the database connection needs to come from the one defined in the config of the webservices application and not from the CMS (https://github.com/fastslack/webservices/blob/newmvc/component/admin/Webservices/Controller/DisplayController.php#L115).

That was one of the reasons I had to hack things up in the existing component (https://github.com/joomla-projects/webservices/blob/develop/component/admin/models/webservice.php#L86-L97)

**Comment by fastslack** [comment]
Nice! thanks for the information. What do you think is the best way to replace the database connection configuration?

The main issue that i'm having right now is that i have not any model class for forms like JModelForm. Do you know any example using forms and new mvc?

**Comment by fastslack** [comment]
https://github.com/fastslack/webservices/commit/b4975f9654a4bb5b6d7caf7ff8be4fbd24688507

**Comment by wilsonge** [comment]
I'm pretty sure there aren't any - largely because the frameworks's form package just never really worked. What exactly isn't working - I think the code in the JModelForm package for dealing with forms is pretty ok

**Comment by fastslack** [comment]
Here a new PR for new MVC move, any comments welcome: https://github.com/joomla-projects/webservices/pull/22

**Comment by fastslack** [comment]
The problem with JModelForm is the getState compatibility that cannot allow to extend it. That class still depending to legacy.

**Pull request merged**